### ISSUE TITLE
read_str() should return a Result instead of panicking

### ISSUE THREAD

**Issue by kevinaboos** [nrc/zero#5]
Wasn't sure whether to file this as an issue in this crate or in your xmas-elf crate, but I chose here because the lowest-level change should be made here. 

In the [`read_str()`](https://github.com/nrc/zero/blob/master/src/lib.rs#L67) function, there is the potential to panic on the `expect` statement. When parsing a file (such as an ELF file using your xmas-elf crate), some symbols may not be encoded as UTF-8, so it'd be nice to be able to skip them and recover gracefully through a returned `Result`, rather than simply panicking and aborting the entire parse thread. 

If you're okay with that change, I'm happy to submit a pull request to both this and the xmas-elf crate. 

Also, thanks a lot for this and the xmas-elf crate, and also for making the change for my previous issue with the elf flags. It's really saved me a lot of time!!

**Comment by nrc** [comment]
We should probably provide both panicking and Result-returning versions of the function - maybe change `read_str` to `expect_str` and add a result-returning `read_str`?

> If you're okay with that change, I'm happy to submit a pull request to both this and the xmas-elf crate.

That would be great thanks!

**Comment by kevinaboos** [comment]
In my forks, I have transformed it to Result-only since that's what I needed. I can submit a pull request and then someone else could add back in / create the expect versions. Or if I have time tomorrow I'll just add back in the panicking version too. 

The one thing I don't love about my version is that is returns a Result where the Err type is `&'static str` instead of `Utf8Error`, which would be preferable imo. You can't create a Utf8Error object, so I couldn't figure out an easy way to return one.

**Comment by nrc** [comment]
I'd prefer not to make a breaking change if we could avoid it (I realise my earlier proposal is breaking, but perhaps there is a naming scheme which is not).

If we're going to start using `Result` we should do it properly and introduce our own error type and use that rather than a string.

**Comment by kevinaboos** [comment]
Agreed. Quite busy this week, but once I get a spare moment I can refactor my forks into a non-breaking naming scheme, and then submit a pull request for each.
